### PR TITLE
Add Tributary Azure Blob snapshot storage provider (vfe agent) +semver: feature

### PR DESCRIPTION
**Pull request by BenjaminLGibbs**
# Add Tributary Azure Blob snapshot storage provider

Adds a first-class Azure Blob-backed snapshot storage provider for Tributary and switches the Spring sample to use Blob snapshots while keeping Brooks event streams on Cosmos.

## Business Value

This gives Mississippi applications a lower-friction snapshot storage option for Azure-hosted workloads that already use Blob Storage. Teams can keep event streams in Cosmos while storing snapshots in a dedicated Blob container, reducing coupling between event persistence and snapshot persistence.

**Key benefits:**

1. Adds an Azure Blob snapshot provider that follows the existing Tributary storage-provider seam.
2. Uses a JSON document format with schema metadata, opaque payload bytes, and optional gzip compression.
3. Demonstrates the provider in Spring using the existing Aspire Blob resource with a dedicated `snapshots` container.

## Common Use Cases

| Domain | Use Case | Example |
| --- | --- | --- |
| SaaS platforms | Separate event stream and snapshot persistence | Keep Brooks events in Cosmos while storing aggregate snapshots in Blob Storage. |
| Cost-sensitive services | Use Blob for larger snapshot payloads | Store infrequently read snapshot JSON documents in a dedicated Blob container. |
| Azure-hosted samples | Reuse an existing storage account safely | Spring reuses the Aspire `blobs` client but isolates snapshots in the `snapshots` container. |

### Overview

The provider registers against the existing `ISnapshotStorageProvider` abstraction. Snapshot writes create one JSON Blob per snapshot version under a hashed stream prefix, and reads validate the persisted document before returning a `SnapshotEnvelope`.

```text
Spring.Runtime
  ├─ Brooks events -> Cosmos provider (unchanged)
  └─ Tributary snapshots -> Blob provider
        ├─ keyed BlobServiceClient: mississippi-blob-snapshots
        ├─ dedicated container: snapshots
        ├─ JSON document wrapper: schemaVersion, stream key, version, compression, payload
        └─ optional gzip payload compression
```

### Key Design Decisions

- **JSON wrapper over opaque payload bytes**: preserves the existing snapshot envelope contract while making storage inspectable and versioned.
- **Dedicated `snapshots` container**: allows one Blob account to serve multiple concerns without mixing snapshot objects with other Blob usage.
- **Hashed stream prefixes**: avoids leaking raw key components into Blob paths and keeps names safe for Azure Blob naming rules.
- **Startup initialization**: a hosted service creates the container if it does not already exist.
- **Bounded gzip decompression**: protects reads from expanding beyond the configured maximum uncompressed payload size.

### Code Example

```csharp
builder.Services.AddBlobSnapshotStorageProvider(options =>
{
    options.EnableCompression = true;
});
```

Spring also forwards the Aspire `blobs` client to `SnapshotBlobDefaults.BlobServiceClientServiceKey` so the provider can reuse the same storage account with its own `snapshots` container.

## Observability

Spring snapshot observability changes from the Cosmos snapshot meter to the Blob snapshot meter.

| Metric/Log | Type | Description |
| --- | --- | --- |
| `Mississippi.Storage.Blob.Snapshots` | Meter | Blob snapshot delete, prune, read, write, size, and duration metrics. |
| Provider logger extensions | Logs | Source-generated diagnostics for provider, repository, and startup initialization behavior. |

Dashboard owners should update any Spring snapshot dashboards that referenced `Mississippi.Storage.Snapshots` to use `Mississippi.Storage.Blob.Snapshots`.

### New Files

| File | Purpose |
| --- | --- |
| `docs/Docusaurus/docs/tributary/storage-providers/blob-storage.md` | Documents Blob snapshot provider registration, storage format, compression, lifecycle, diagnostics, and Spring usage. |
| `src/Tributary.Runtime.Storage.Blobs/Tributary.Runtime.Storage.Blobs.csproj` | New provider project. |
| `src/Tributary.Runtime.Storage.Blobs/Diagnostics/SnapshotBlobStorageMetrics.cs` | Blob snapshot metrics. |
| `src/Tributary.Runtime.Storage.Blobs/ISnapshotBlobOperations.cs` | Azure SDK seam for deterministic tests. |
| `src/Tributary.Runtime.Storage.Blobs/ISnapshotBlobRepository.cs` | Internal repository seam. |
| `src/Tributary.Runtime.Storage.Blobs/SnapshotBlobContainerInitializer.cs` | Startup container initialization hosted service. |
| `src/Tributary.Runtime.Storage.Blobs/SnapshotBlobContainerInitializerLoggerExtensions.cs` | Source-generated initializer logging. |
| `src/Tributary.Runtime.Storage.Blobs/SnapshotBlobDefaults.cs` | Provider constants, service keys, default container, and default payload limit. |
| `src/Tributary.Runtime.Storage.Blobs/SnapshotBlobStorageOptions.cs` | Provider options. |
| `src/Tributary.Runtime.Storage.Blobs/SnapshotBlobStorageOptionsValidator.cs` | Startup options validation. |
| `src/Tributary.Runtime.Storage.Blobs/SnapshotBlobStorageProvider.cs` | Public snapshot storage provider implementation. |
| `src/Tributary.Runtime.Storage.Blobs/SnapshotBlobStorageProviderLoggerExtensions.cs` | Source-generated provider logging. |
| `src/Tributary.Runtime.Storage.Blobs/SnapshotBlobStorageProviderRegistrations.cs` | DI registration extensions and keyed Blob clients. |
| `src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobCompression.cs` | Optional gzip compression and bounded decompression. |
| `src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobCompressionResult.cs` | Compression result value object. |
| `src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobDocument.cs` | JSON document schema. |
| `src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobDocumentSerializer.cs` | JSON serialization/deserialization. |
| `src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobOperations.cs` | Azure Blob SDK adapter. |
| `src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobPath.cs` | Durable Blob path construction and version parsing. |
| `src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobRepository.cs` | Read/write/delete/prune semantics and validation. |
| `src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobRepositoryLoggerExtensions.cs` | Source-generated repository logging. |
| `src/Tributary.Runtime.Storage.Blobs/packages.lock.json` | Lock file for the provider project. |
| `tests/Tributary.Runtime.Storage.Blobs.L0Tests/Tributary.Runtime.Storage.Blobs.L0Tests.csproj` | New L0 test project. |
| `tests/Tributary.Runtime.Storage.Blobs.L0Tests/Diagnostics/SnapshotBlobStorageMetricsTests.cs` | Metrics tests. |
| `tests/Tributary.Runtime.Storage.Blobs.L0Tests/SnapshotBlobCompressionTests.cs` | Compression tests. |
| `tests/Tributary.Runtime.Storage.Blobs.L0Tests/SnapshotBlobDocumentTests.cs` | JSON document serialization tests. |
| `tests/Tributary.Runtime.Storage.Blobs.L0Tests/SnapshotBlobNameTests.cs` | Blob naming/path tests. |
| `tests/Tributary.Runtime.Storage.Blobs.L0Tests/SnapshotBlobRepositoryTests.cs` | Repository behavior and validation tests. |
| `tests/Tributary.Runtime.Storage.Blobs.L0Tests/SnapshotBlobStorageOptionsTests.cs` | Defaults and options validation tests. |
| `tests/Tributary.Runtime.Storage.Blobs.L0Tests/SnapshotBlobStorageProviderRegistrationsTests.cs` | DI registration tests. |
| `tests/Tributary.Runtime.Storage.Blobs.L0Tests/SnapshotBlobStorageProviderTests.cs` | Provider delegation tests. |
| `tests/Tributary.Runtime.Storage.Blobs.L0Tests/packages.lock.json` | Lock file for the test project. |

### Modified Files

| File | Change |
| --- | --- |
| `docs/Docusaurus/docs/tributary/storage-providers/index.md` | Adds Azure Blob snapshots to the storage provider overview and notes Spring’s split between Blob snapshots and Cosmos events. |
| `mississippi.slnx` | Includes the new provider and L0 test projects. |
| `samples/Spring/Spring.AppHost/Program.cs` | Clarifies existing Blob resource comments for locking and snapshots. |
| `samples/Spring/Spring.Runtime/Program.cs` | Switches Spring snapshots to Blob, enables compression, forwards the keyed Blob client, and registers Blob snapshot metrics. |
| `samples/Spring/Spring.Runtime/Spring.Runtime.csproj` | References the new Blob provider project. |
| `samples/Spring/Spring.Runtime/packages.lock.json` | Refreshes lock file after adding the project reference. |
| `tests/Architecture.L0Tests/Architecture.L0Tests.csproj` | Adds the provider assembly to architecture validation. |
| `tests/Architecture.L0Tests/packages.lock.json` | Refreshes architecture test lock file. |

### New Tests

| File | Coverage |
| --- | --- |
| `tests/Tributary.Runtime.Storage.Blobs.L0Tests/**` | 51 deterministic L0 tests covering registration, options, paths, JSON schema, compression, repository validation, provider delegation, and metrics. |

## Quality Gates

- [x] Blob provider L0 tests passed: `51/51`.
- [x] Spring runtime build passed with warnings as errors.
- [x] Architecture L0 tests passed: `36/36`.
- [x] Docusaurus production build passed.
- [x] Full `pwsh .\go.ps1 -IncludeMutation` completed successfully with exit code `0`.
- [x] Final zero-warning builds passed for `mississippi.slnx` and `samples.slnx`.

Note: the mutation phase emitted broad Stryker project-level warnings while repository automation still completed successfully, reported acceptable mutation quality, and the full pipeline exited `0`.

## Migration Notes

No breaking framework API change is intended.

Spring sample behavior changes:

- Snapshot storage moves from Cosmos snapshots to Azure Blob snapshots.
- Brooks event stream storage remains on Cosmos.
- Snapshot metrics now use `Mississippi.Storage.Blob.Snapshots` instead of `Mississippi.Storage.Snapshots`.

Existing applications can opt into the provider by registering a keyed `BlobServiceClient` and calling `AddBlobSnapshotStorageProvider`.

## Related Issues

N/A

### CODE REVIEW

File: src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobCompression.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+                output.Write(buffer, 0, bytesRead);
+            }
+
+            return output.ToArray();
+        }
+        catch (InvalidDataException exception)
+        {
+            throw new InvalidDataException("Stored gzip snapshot payload is invalid.", exception);
+        }
```
[the remark this answers is not in the document]
Comment (BenjaminLGibbs) [reply src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobCompression.cs]: Fixed in 4e0bf0d6 (included in current head 90935497). Removed the broad gzip exception wrapper so a payload that exceeds the configured limit retains its specific diagnostic. Regression coverage checks that message, and corrupt gzip still raises InvalidDataException. The final provider suite passes 174 tests with warnings as errors; full cleanup and the full build/test pipeline also passed.

File: src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobRepository.cs [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        try
+        {
+            return Convert.FromBase64String(base64Payload);
+        }
+        catch (FormatException exception)
+        {
+            throw new InvalidDataException(
+                $"Blob snapshot '{blobName}' contains invalid Base64 payload data.",
+                exception);
+        }
```
[the remark this answers is not in the document]
Comment (BenjaminLGibbs) [reply src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobRepository.cs]: Fixed in 843f712b and hardened further in c03f0f33 (current head 90935497). The JSON boundary now requires every schema field and rejects null string metadata before decoding; malformed persisted payloads consistently raise InvalidDataException and receive contextual repository logging. Raw-JSON tests cover null data and every missing/null schema property. All 174 provider tests pass with warnings as errors.

File: src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobCompression.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    /// <returns>The uncompressed payload bytes.</returns>
+    public static byte[] Decompress(
+        string compression,
+        byte[] storedBytes,
+        long maximumPayloadSizeBytes = long.MaxValue
+    )
+    {
+        ArgumentNullException.ThrowIfNull(compression);
+        ArgumentNullException.ThrowIfNull(storedBytes);
+        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(maximumPayloadSizeBytes);
```
[the remark this answers is not in the document]
Comment (BenjaminLGibbs) [reply src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobCompression.cs]: Fixed in 0341e5c9 and the strict-schema codec change c03f0f33 (current head 90935497). Null persisted compression metadata now raises InvalidDataException at deserialization; the decompression helper also treats null compression as invalid data. Tests cover both entry points, and the repository logs invalid-document failures consistently. All 174 provider tests pass with warnings as errors.

File: src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobRepository.cs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        Logger.DeletingAllSnapshots(prefix);
+        await foreach (string blobName in Operations.ListBlobNamesAsync(prefix, cancellationToken))
+        {
+            if (!SnapshotBlobPath.TryParseVersionFromBlobName(blobName, streamKey, out long _))
+            {
+                continue;
+            }
+
+            await Operations.DeleteIfExistsAsync(blobName, cancellationToken).ConfigureAwait(false);
+            Logger.SnapshotDeleted(blobName);
```
[the remark this answers is not in the document]
Comment (BenjaminLGibbs) [reply src/Tributary.Runtime.Storage.Blobs/Storage/SnapshotBlobRepository.cs]: Fixed in a248be95. Both DeleteAllAsync and DeleteAsync now emit the repository's SnapshotDeleted event only when DeleteIfExistsAsync returns true. The provider-level messages describe deletion completion so an idempotent request for a missing Blob does not claim a removal. Four regression cases cover existing/missing Blobs through both paths. All 178 provider tests pass with warnings as errors; targeted cleanup also completed.

File: src/Tributary.Runtime.Storage.Blobs/Tributary.Runtime.Storage.Blobs.csproj [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
[the remark this answers is not in the document]
Comment (BenjaminLGibbs) [reply src/Tributary.Runtime.Storage.Blobs/Tributary.Runtime.Storage.Blobs.csproj]: The author has explicitly directed: “Refuse to split it. Keep it as one.” I have recorded this PR-specific scope decision in the description; the general repository review rule is unchanged. The feature was reviewed in separate persistence/format, registration/diagnostics, and integration/verification areas, including independent source review and passing local build/test, cleanup, and documentation gates. The suggested provider/sample/docs split would still leave 4,983 handwritten provider/test lines plus generated locks, so it would not satisfy the stated size threshold. Keeping this cohesive feature in PR #484 and resolving the process request according to the author's decision.

File: tests/Tributary.Runtime.Storage.Blobs.L0Tests/SnapshotBlobStorageOptionsTests.cs [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        {
+            ValidateOptionsResult result = validator.Validate(
+                Options.DefaultName,
+                new()
+                {
+                    MaximumSnapshotPayloadSizeBytes = limit,
+                    MaximumSnapshotDocumentSizeBytes = limit,
+                });
+            Assert.True(result.Succeeded);
+        }
```
[the remark this answers is not in the document]
Comment (BenjaminLGibbs) [reply tests/Tributary.Runtime.Storage.Blobs.L0Tests/SnapshotBlobStorageOptionsTests.cs]: Updated in 575e7ac2. The boundary values are now explicitly mapped to ValidateOptionsResult instances with Select, and the loop only asserts those results. Targeted cleanup completed and all 178 provider tests pass with warnings as errors.
